### PR TITLE
[CLI/logs-api] Update logs cli / api.

### PR DESCRIPTION
**Pull request by charlesg3**
**Please See** [Git Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions)

*Issue #, if available:*

*Description of changes:*
- [CLI/logs-api] Update logs cli / api.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: api/spec/openapi/ParallelCluster.openapi.yaml [hunk 1]
```diff
@@ -999,6 +1252,240 @@ paths:
         type: aws_proxy
         uri:
           Fn::Sub: arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${ParallelClusterFunction.Arn}/invocations
+  /v3/images/custom/{imageId}/logstreams:
+    get:
+      description: Retrieve the list of log streams associated with a cluster.
```
Comment (enrico-usai) [inline api/spec/openapi/ParallelCluster.openapi.yaml:1257]: cluster --> image

File: api/spec/openapi/ParallelCluster.openapi.yaml [hunk 2]
```diff
@@ -999,6 +1252,240 @@ paths:
...
+            type: string
+            pattern: ^[a-zA-Z][a-zA-Z0-9-]+$
+            description: Id of the image.
+          required: true
+        - name: region
+          in: query
+          description: Region that the given cluster belongs to.
+          schema:
+            type: string
+            description: Region that the given cluster belongs to.
```
Comment (enrico-usai) [inline api/spec/openapi/ParallelCluster.openapi.yaml:1273]: cluster --> image

File: api/spec/openapi/ParallelCluster.openapi.yaml [hunk 3]
```diff
@@ -697,6 +697,259 @@ paths:
...
+        - Cluster Logs
+      x-amazon-apigateway-integration:
+        credentials:
+          Fn::Sub: ${APIGatewayExecutionRole.Arn}
+        httpMethod: POST
+        payloadFormatVersion: "2.0"
+        type: aws_proxy
+        uri:
+          Fn::Sub: arn:${AWS::Partition}:apigateway:${AWS::Region}:lambda:path/2015-03-31/functions/${ParallelClusterFunction.Arn}/invocations
+  /v3/clusters/{clusterName}/stackevents:
```
Comment (enrico-usai) [inline api/spec/openapi/ParallelCluster.openapi.yaml:887]: In future we can have more than 1 stack. How can we list the events of substacks?
Should we expose a stacks and stacks/{stackName}/events ?

File: api/spec/smithy/model/operations/ListImageLogStreams.smithy [hunk 4]
```diff
@@ -0,0 +1,31 @@
+namespace parallelcluster
+
+@suppress(["InputOutputStructureReuse"])
+@paginated
+@readonly
+@http(method: "GET", uri: "/v3/images/custom/{imageId}/logstreams", code: 200)
+@tags(["Image Logs"])
+@documentation("Retrieve the list of log streams associated with a cluster.")
```
Comment (enrico-usai) [inline api/spec/smithy/model/operations/ListImageLogStreams.smithy:8]: cluster --> image

File: api/spec/smithy/model/operations/ListImageLogStreams.smithy [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+        InternalServiceException,
+        BadRequestException,
+        UnauthorizedClientError,
+        LimitExceededException,
+    ]
+}
+
+structure ListImageLogStreamsRequest {
+    @httpQuery("region")
+    @documentation("Region that the given cluster belongs to.")
```
Comment (enrico-usai) [inline api/spec/smithy/model/operations/ListImageLogStreams.smithy:22]: same here

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    :param next_token: Token to use for paginated requests.
+    :type next_token: str
+
+    :rtype: GetClusterStackEventsResponseContent
+    """
+    cluster = Cluster(cluster_name)
+    stack_events = cluster.get_stack_events(next_token=next_token)
+
+    def convert_event(event):
+        event = {k[0].lower() + k[1:]: v for k, v in event.items()}
```
Comment (enrico-usai) [inline cli/src/pcluster/api/controllers/cluster_logs_controller.py:145]: why do we need to lowerize the first character of the key?

File: cli/src/pcluster/api/controllers/image_logs_controller.py [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    events = [convert_event(event) for event in stack_events["StackEvents"]]
+    return GetImageStackEventsResponseContent(next_token=stack_events.get("NextToken", None), events=events)
+
+
+@configure_aws_region()
+@convert_errors()
+def list_image_log_streams(image_id, region=None, next_token=None):
+    """
+    Retrieve the list of log streams associated with a cluster.
```
Comment (enrico-usai) [inline cli/src/pcluster/api/controllers/image_logs_controller.py:132]: cluster --> image

File: cli/src/pcluster/api/models/cloud_formation_stack_status.py [hunk 8]
```diff
@@ -1,13 +1,9 @@
...
-#
-# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
-# with the License. A copy of the License is located at http://aws.amazon.com/apache2.0/
-# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
-# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
-# limitations under the License.
+# coding: utf-8
 
-# pylint: disable=R0801
+from __future__ import absolute_import
```
Comment (enrico-usai) [inline cli/src/pcluster/api/models/cloud_formation_stack_status.py:3]: I don't think we need future import. It's for python 2 compatibility.
Is this auto-generated?

File: cli/src/pcluster/utils.py [hunk 9]
```diff
@@ -118,31 +118,45 @@ def get_supported_os_for_architecture(architecture):
...
 
+def to_camel_case(snake_case_word):
+    """Convert the given snake case word into a camelCase one."""
+    parts = iter(snake_case_word.split("_"))
+    pascal = "".join(word.title() for word in parts)
+    return pascal[0].lower() + pascal[1:]
+
+
+def to_pascal_case(snake_case_word):
+    """Convert the given snake case word into a PamelCase one."""
```
Comment (enrico-usai) [inline cli/src/pcluster/utils.py:155]: PamelCase :D

File: cli/src/pcluster/utils.py [hunk 10]
```diff
@@ -118,31 +118,45 @@ def get_supported_os_for_architecture(architecture):
...
+            raise TypeError("to_iso_time object must be 'str', 'int' or 'datetime'.")
         return time_.isoformat(timespec="milliseconds")[:-6] + "Z"
     else:
         return time_in
 
 
+def to_camel_case(snake_case_word):
+    """Convert the given snake case word into a camelCase one."""
+    parts = iter(snake_case_word.split("_"))
+    pascal = "".join(word.title() for word in parts)
```
Comment (enrico-usai) [inline cli/src/pcluster/utils.py:150]: what about reusing the to_pascal_case function?

File: cli/tests/pcluster/cli/test_get_cluster_log_events.py [hunk 11]
```diff
@@ -40,11 +42,10 @@ def test_required_args(self, args, error_message, run_cli, capsys):
...
         [
-            ({"start_time": "wrong"}, "Start time and end time filters must be in the ISO 8601 format"),
-            ({"end_time": "1622802790248"}, "Start time and end time filters must be in the ISO 8601 format"),
-            ({"head": "wrong"}, "invalid int value"),
-            ({"tail": "wrong"}, "invalid int value"),
-            ({"stream_period": "wrong"}, "invalid int value"),
+            ({"start_time": "wrong"}, "start_time filter must be in the ISO 8601 format"),
+            ({"end_time": "1622802790248"}, "end_time filter must be in the ISO 8601 format"),
+            ({"start_from_head": "wrong"}, "expected 'boolean' for parameter 'start-from-head'"),
+            ({"limit": "wrong"}, "expected 'number' for parameter 'limit'"),
```
Comment (enrico-usai) [inline cli/tests/pcluster/cli/test_get_cluster_log_events.py:48]: let's update the changelog to remove stream, stream-period, other removed parameters and double check the logs commands

File: cli/tests/pcluster/cli/test_get_cluster_log_events.py [hunk 12]
```diff
@@ -116,68 +110,34 @@ def test_execute(self, mocker, capsys, set_env, run_cli, test_datadir, assert_ou
...
-        if "tail" in args:
-            cli_args.extend(["--tail", args["tail"]])
-        if "next_token" in args:
-            cli_args.extend(["--next-token", args["next_token"]])
-        if "stream" in args:
-            cli_args.extend(["--stream"])
-        if "stream_period" in args:
-            cli_args.extend(["--stream-period", args["stream_period"]])
+        for k, val in args.items():
+            cli_args.extend([f"--{to_kebab_case(k)}", val])
```
Comment (enrico-usai) [inline cli/tests/pcluster/cli/test_get_cluster_log_events.py:142]: nice idea!

File: cli/tests/pcluster/cli/test_list_image_log_streams/TestListImageLogStreamsCommand/test_helper/pcluster-help.txt [hunk 13]
```diff
@@ -0,0 +1,15 @@
+usage: pcluster list-image-log-streams [-h] --image-id IMAGE_ID
+                                       [--region REGION]
+                                       [--next-token NEXT_TOKEN] [--debug]
+                                       [--query QUERY]
+
+Retrieve the list of log streams associated with a cluster.
```
Comment (enrico-usai) [inline cli/tests/pcluster/cli/test_list_image_log_streams/TestListImageLogStreamsCommand/test_helper/pcluster-help.txt:6]: cluster --> image

File: api/spec/smithy/model/operations/GetClusterLogEvents.smithy [hunk 14]
```diff
@@ -0,0 +1,43 @@
+namespace parallelcluster
+
+@suppress(["MissingPaginatedTrait", "PaginatedTrait", "InputOutputStructureReuse"])
```
Comment (demartinofra) [inline api/spec/smithy/model/operations/GetClusterLogEvents.smithy:3]: why do we need to suppress "MissingPaginatedTrait", "PaginatedTrait"? isn't this properly paginated?

for the ones ew need to suppress can you leave a comment in the code explaining why.

Same for all other operations in this PR

File: api/spec/smithy/model/operations/GetClusterLogEvents.smithy [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+@paginated
+@readonly
+@http(method: "GET", uri: "/v3/clusters/{clusterName}/logstreams/{logStreamName}", code: 200)
+@tags(["Cluster Logs"])
+@documentation("Retrieve the events associated with a log stream.")
+operation GetClusterLogEvents {
+    input: GetClusterLogEventsRequest,
+    output: GetLogEventsResponse,
+    errors: [
```
Comment (demartinofra) [inline api/spec/smithy/model/operations/GetClusterLogEvents.smithy:12]: we probably also need NotFoundException and in the message we should discriminate between cluster not found and log stream not found

same for image

File: api/spec/smithy/model/operations/GetClusterStackEvents.smithy [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+@suppress(["MissingPaginatedTrait", "PaginatedTrait", "InputOutputStructureReuse", "ResourceIdentifierBinding"])
+@paginated
+@readonly
+@http(method: "GET", uri: "/v3/clusters/{clusterName}/stackevents", code: 200)
+@tags(["Cluster Logs"])
+@documentation("Retrieve the events associated with the stack for a given cluster.")
+operation GetClusterStackEvents {
+    input: GetClusterStackEventsRequest,
+    output: GetStackEventsResponse,
+    errors: [
```
Comment (demartinofra) [inline api/spec/smithy/model/operations/GetClusterStackEvents.smithy:12]: missing NotFoundException. Same for image

File: api/spec/smithy/model/operations/ListClusterLogStreams.smithy [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+@suppress(["InputOutputStructureReuse"])
+@paginated
+@readonly
+@http(method: "GET", uri: "/v3/clusters/{clusterName}/logstreams", code: 200)
+@tags(["Cluster Logs"])
+@documentation("Retrieve the list of log streams associated with a cluster.")
+operation ListClusterLogStreams {
+    input: ListClusterLogStreamsRequest,
+    output: ListLogStreamsResponse,
+    errors: [
```
Comment (demartinofra) [inline api/spec/smithy/model/operations/ListClusterLogStreams.smithy:12]: NotFoundException, and same for image

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py (hunk 6 above)
Comment (charlesg3) [reply cli/src/pcluster/api/controllers/cluster_logs_controller.py:145]: This is the format that the generated model expects when it's creating the response object (vs what is provided from AWS) so we are just translating between the two.

File: cli/src/pcluster/api/models/cloud_formation_stack_status.py (hunk 8 above)
Comment (charlesg3) [reply cli/src/pcluster/api/models/cloud_formation_stack_status.py:3]: Yes this is auto generated.

File: cli/src/pcluster/utils.py (hunk 10 above)
Comment (charlesg3) [reply cli/src/pcluster/utils.py:150]: Good call.

File: cli/src/pcluster/utils.py (hunk 9 above)
Comment (charlesg3) [reply cli/src/pcluster/utils.py:155]: :camel:  vs :dromedary_camel:

File: cli/tests/pcluster/cli/test_get_cluster_log_events.py (hunk 11 above)
Comment (charlesg3) [reply cli/tests/pcluster/cli/test_get_cluster_log_events.py:48]: Can you clarify what you mean by double checking the logs commands?

File: cli/tests/pcluster/cli/test_get_cluster_log_events.py (hunk 12 above)
Comment (charlesg3) [reply cli/tests/pcluster/cli/test_get_cluster_log_events.py:142]: :raised_hands:

**Review by enrico-usai — APPROVED** [review]

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py [hunk 18]
```diff
@@ -0,0 +1,194 @@
...
+        if not check_cluster_version(cluster):
+            raise BadRequestException(
+                f"cluster '{cluster_name}' belongs to an incompatible ParallelCluster major version."
+            )
+    except StackNotFoundError:
+        raise NotFoundException(
+            f"cluster '{cluster_name}' does not exist or belongs to an incompatible ParallelCluster major version."
+        )
+    log_events = cluster.get_log_events(
+        log_stream_name,
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/cluster_logs_controller.py:107]: what happens if the log group or the log stream is not found?

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py [hunk 19]
```diff
@@ -0,0 +1,194 @@
...
+    :type cluster_name: str
+    :param region: AWS Region that the operation corresponds to.
+    :type region: str
+    :param next_token: Token to use for paginated requests.
+    :type next_token: str
+
+    :rtype: GetClusterStackEventsResponseContent
+    """
+    cluster = Cluster(cluster_name)
+    stack_events = cluster.get_stack_events(next_token=next_token)
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/cluster_logs_controller.py:142]: what if cluster does not exist?

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py [hunk 20]
```diff
@@ -0,0 +1,194 @@
...
+    cluster = Cluster(cluster_name)
+    try:
+        if not check_cluster_version(cluster):
+            raise BadRequestException(
+                f"cluster '{cluster_name}' belongs to an incompatible ParallelCluster major version."
+            )
+    except StackNotFoundError:
+        raise NotFoundException(
+            f"cluster '{cluster_name}' does not exist or belongs to an incompatible ParallelCluster major version. "
+        )
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/cluster_logs_controller.py:181]: minor: this is almost everywhere and we could export to a function

File: cli/src/pcluster/api/controllers/cluster_logs_controller.py [hunk 21]
```diff
@@ -0,0 +1,194 @@
...
+
+    def convert_log(log):
+        log["logStreamArn"] = log.pop("arn")
+        if "storedBytes" in log:
+            del log["storedBytes"]
+        for ts_name in ["creationTime", "firstEventTimestamp", "lastEventTimestamp", "lastIngestionTime"]:
+            log[ts_name] = to_iso_time(log[ts_name])
+        return LogStream.from_dict(log)
+
+    cluster_logs = cluster.list_logs(filters=filters, next_token=next_token)
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/cluster_logs_controller.py:191]: are we handling the case when the CW log groups do not exist by returning an empty array?

File: cli/src/pcluster/api/controllers/image_logs_controller.py [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+    ListImageLogStreamsResponseContent,
+    LogEvent,
+    LogStream,
+    StackEvent,
+)
+from pcluster.models.imagebuilder import ImageBuilder
+from pcluster.utils import to_iso_time
+
+
+def _validate_timestamp(val, ts_name):
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/image_logs_controller.py:26]: minor: can be put in common and reused in both controllers

File: cli/src/pcluster/api/controllers/image_logs_controller.py [hunk 23]
```diff
@@ -0,0 +1,156 @@
...
+
+    :rtype: GetImageLogEventsResponseContent
+    """
+    if start_time:
+        _validate_timestamp(start_time, "start_time")
+    if end_time:
+        _validate_timestamp(end_time, "end_time")
+
+    imagebuilder = ImageBuilder(image_id=image_id)
+    log_events = imagebuilder.get_log_events(
```
Comment (demartinofra) [inline cli/src/pcluster/api/controllers/image_logs_controller.py:81]: what if image does not exist? I'm afraid in the model we are always throwing a generic ClusterActionError which gets translated into an InternalServiceException for the API. I believe we need to use more appropriate exceptions to have the correct status code from the API

Comment (charlesg3) [reply cli/src/pcluster/api/controllers/image_logs_controller.py:81]: See comment on PR

**Comment by charlesg3** [comment]
https://github.com/aws/aws-parallelcluster/pull/3064 :point_left: Addresses concerns brought up by @demartinofra (now marked as resolved)

**Pull request merged**